cumulus/pallets/parachain-system/src/lib.rs

```diff
@@ -1171,33 +1180,38 @@ impl<T: Config> Pallet<T> {
 	/// hash doesn't match the expected.
 	fn enqueue_inbound_downward_messages(
 		expected_dmq_mqc_head: relay_chain::Hash,
-		downward_messages: Vec<InboundDownwardMessage>,
+		downward_messages: CompressedInboundDownwardMessages,
 	) -> Weight {
-		let dm_count = downward_messages.len() as u32;
+		downward_messages.check_advancement_rule("DMQ");
+
 		let mut dmq_head = <LastDmqMqcHead<T>>::get();
 
-		let weight_used = T::WeightInfo::enqueue_inbound_downward_messages(dm_count);
-		if dm_count != 0 {
-			Self::deposit_event(Event::DownwardMessagesReceived { count: dm_count });
+		let (messages, hashed_messages) = downward_messages.messages();
+		let message_count = messages.len() as u32;
+		let weight_used = T::WeightInfo::enqueue_inbound_downward_messages(message_count);
+		if message_count != 0 {
+			Self::deposit_event(Event::DownwardMessagesReceived { count: message_count });
 
 			// Eagerly update the MQC head hash:
-			for m in &downward_messages {
-				dmq_head.extend_downward(m);
+			let mut last_processed_msg = InboundMessageId { sent_at: 0, reverse_idx: 0 };
+			for msg in messages {
+				dmq_head.extend_downward(msg);
```

Comment:
If we could not include messages in the previous block, the dmq head would already have "included" them. This means you will apply them a second time and the head will be wrong.

We should keep the intermediate dmq head and just calculate the final one to compare against the one stored at the relay chain.